src/media-server.js

```diff
@@ -0,0 +1,199 @@
+import { once } from 'events'
+import { promisify } from 'util'
+import fastify from 'fastify'
+import pTimeout from 'p-timeout'
+import StateMachine from 'start-stop-state-machine'
+
+import BlobServerPlugin from './fastify-plugins/blobs.js'
+import { kBlobStore } from './mapeo-project.js'
+
+export const BLOBS_PREFIX = 'blobs'
+export const ICONS_PREFIX = 'icons'
+
+/**
+ * @typedef {Object} StartOpts
+ *
+ * @property {string} [host]
+ * @property {number} [port]
+ */
+
+export class MediaServer {
+  #fastify
+  #fastifyStarted
+  #host
+  #port
+  #serverState
+
+  /**
+   * @param {object} params
+   * @param {(projectPublicId: string) => Promise<import('./mapeo-project.js').MapeoProject>} params.getProject
+   * @param {import('fastify').FastifyServerOptions['logger']} [params.logger]
+   */
+  constructor({ getProject, logger }) {
+    this.#fastifyStarted = false
+    this.#host = '127.0.0.1'
+    this.#port = 0
+
+    this.#fastify = fastify({ logger })
+
+    // Don't accept new connections when closing/closed
+    this.#fastify.addHook('onRequest', this.#onRequestHook.bind(this))
+
+    this.#fastify.register(BlobServerPlugin, {
+      prefix: BLOBS_PREFIX,
+      getBlobStore: async (projectPublicId) => {
+        const project = await getProject(projectPublicId)
+        return project[kBlobStore]
+      },
+    })
+
+    this.#serverState = new StateMachine({
+      start: this.#startServer.bind(this),
+      stop: this.#stopServer.bind(this),
+    })
+  }
+
+  /**
+   * This is necessary because a keep-alive connection from another device will
+   * prevent this server from closing. This hook ensures that if this server is
+   * in the "stopping", "stopped" or "error" states, then it responds with the
+   * "Connection: close" header, which tells the keep-alive client to stop. It
+   * also responds with a 503 "Service unavailable" error.
+   *
+   * @param {import('fastify').FastifyRequest} request
+   * @param {import('fastify').FastifyReply} reply
+   */
+  async #onRequestHook(request, reply) {
```

Comment:
I don't fully understand how this can be called, because we call `server.close()` in the same tick as changing state to `stopping`, which [stops the server accepting new connections](https://nodejs.org/api/net.html#net_server_close_callback), so I would have thought we would not receive a request when in the `stopping` or `stopped` states, but I don't fully understand how keep-alive connections work and how `server.close()` interacts with them! Anyway, TL;DR don't understand this enough to comment, but looks smart and looks like it works

Reply:
this was actually plucked from something you implemented in mapeo mobile!

https://github.com/digidem/mapeo-mobile/blob/develop/src/backend/upgrade-manager/upgrade-server.js#L107

I actually added it thinking it could help me figure out the issues I was running into, but turns out it wasn't related. can remove this if you think it's not necessary for now, since I was thinking the same thing (at least for the time being)

Reply:
decided to take out for now (https://github.com/digidem/mapeo-core-next/pull/365/commits/3d4bbd431aed20714cc6837443bd1f24b862e81a). can revisit if we end up seeing notable issues later